OSDOCS-11169: making CSI optional for microshift (#80823)
openshift/openshift-docs: modules/microshift-disabling-lvms-csi-snapshot.adoc

// Module included in the following assemblies:
//
// * microshift_storage/microshift-storage-plugin-overview.adoc
// * microshift_configuring/microshift-disable-lvms-csi-provider-csi-snapshot.adoc

:_mod-docs-content-type: PROCEDURE
[id="microshift-disabling-lvms-csi-snapshot_{context}"]
= Disabling deployments that run CSI snapshot implementations

Before you disable the CSI snapshot implementation pods, you must manually stop any user workloads that have attached volumes. Use the following procedure to disable installation of the CSI implementation pods.

[NOTE]
====
{microshift-short} will not delete CSI snapshot implementation pods. You must configure {microshift-short} to disable installation of the CSI snapshot implementation pods during the startup process.
====

.Procedure

> . To disable installation of the CSI snapshot controller, enter the `optionalCsiComponents` value under the `storage` section of the {microshift-short} configuration file:

Reviewer: ```suggestion
. Disable installation of the CSI snapshot controller by entering the `optionalCsiComponents` value under the `storage` section of the {microshift-short} configuration file:
```

Author: done

+
[source,yaml]
----
# ...
  storage

>    optionalCsiComponents: "csi-snapshot-controller" <1>

Reviewer 2: The value of `csi-snapshot-controller` throws an error when restarting MicroShift.

The `config.yaml/default` entry looks like this

`   # Allowed values are: unset, [], or one or more of ["snapshot-controller", "snapshot-webhook"]

I think `csi` is not needed prepended to the component. I would check with development

Author: @copejon Please confirm

Reviewer 3: correct

Author: done. csi is removed as suggested

>    optionalCsiComponents: "snapshot-controller" <1>

Reviewer 2: This is not constructed properly and will cause a failure when the microshift service is restarted

```suggestion
   optionalCsiComponents: 
        - "snapshot-controller" <1>
```

Author: updated for both "microshift-disabling-lvms-csi-snapshot.adoc" and "microshift-disabling-lvms-csi-driver.adoc"

# ...
----

> <1> Accepted values are `csi-snapshot-controller`, `csi-snapshot-webhook`, or `none`. `none` is mutually exclusive with all other values.

Reviewer 2: see comment above

Author: @copejon Please confirm

Reviewer 3: correct

Author: done. csi is removed as suggested. 
"Unset" is an accepted value?  is this to be added in the list?

Author: refer to the below comment for clarity

+
[NOTE]
====

> If the `optionalCsiComponents` value is empty or null, {microshift-short} defaults to deploying snapshot-controller and snapshot-webhook.

Reviewer 2: Can you ask a developer to provide more details around empty or null ?

I tried 

`   optionalCsiComponents:
       - [""]
`

'   optionalCsiComponents:
       - ""
'

'   optionalCsiComponents:
       - ""
'
Each version fails when I try to restart microshift.service

Author: @copejon Please confirm

Reviewer 3: This is dev bug. I need to confirm a solution before I can comment here.

Reviewer 3: A fix is up, but CI his having a rough day: https://github.com/openshift/microshift/pull/3914

An empty value would be an empty array: `[]`
A null value would be an undefined value: `optionalCsiComponents: ` or simply not providing the field at all:

```
storage: {}
```

Reviewer 3: To summarize, valid entries are:
1. Not defining the field at all, as it's optional.
2. Specifying the field with an empty value ( `[]` )
3. Specifying the field with one of the accepted values

Author: done

====

> . Restart {microshift-short} after the `optionalCsiComponents` field is specified with a supported value in the {microshift-short} configuration file by running the following command:

Reviewer: ```suggestion
. After the `optionalCsiComponents` field is specified with a supported value in the `config.yaml`, restart {microshift-short}  by running the following command:
```
Shortens it a bit.

+
[source,terminal]
----
$ sudo systemctl restart microshift
----
+
[NOTE]
====

> {microshift-short} does not redeploy the disabled components after a restart operation.

Reviewer: ```suggestion
{microshift-short} does not redeploy the disabled components after a restart.
```

Author: done

// Module included in the following assemblies:
//
// * microshift_storage/microshift-storage-plugin-overview.adoc
// * microshift_configuring/microshift-disable-lvms-csi-provider-csi-snapshot.adoc

:_mod-docs-content-type: PROCEDURE
[id="microshift-disabling-lvms-csi-snapshot_{context}"]
= Disabling deployments that run CSI snapshot implementations

> Before you disable the CSI snapshot implementation pods, you must manually stop any user workloads that have attached volumes. Use the following procedure to disable installation of the CSI implementation pods.
> 

Reviewer 3: ```suggestion
```

This is irrelevant to the context. MicroShift will not have been started yet, and thus there will be no workloads running and no volumes to delete.

Author: done

[NOTE]
====
{microshift-short} will not delete CSI snapshot implementation pods. You must configure {microshift-short} to disable installation of the CSI snapshot implementation pods during the startup process.
====

> .Procedure
> 

Reviewer 3: ```suggestion
.Procedure

.Prerequisites
- MicroShift has not be started for the first time
```

Author: done

. Disable installation of the CSI snapshot controller by entering the `optionalCsiComponents` value under the `storage` section of the {microshift-short} configuration file in `/etc/microshift/config.yaml`:
+
[source,yaml]
----
# ...
  storage
   optionalCsiComponents:
    - "snapshot-controller" <1>
# ...
----
<1> Accepted values are `snapshot-controller`, `snapshot-webhook`, or `none`. `none` is mutually exclusive with all other values.
+
[NOTE]
====
If the `optionalCsiComponents` value is empty or null, {microshift-short} defaults to deploying snapshot-controller and snapshot-webhook.
====

> . After the `optionalCsiComponents` field is specified with a supported value in the `config.yaml`, restart {microshift-short} by running the following command:
> +
> [source,terminal]
> ----
> $ sudo systemctl restart microshift
> ----

Reviewer 3: ```suggestion
. After the `optionalCsiComponents` field is specified with a supported value in the `config.yaml`, start {microshift-short} by running the following command:
+
[source,terminal]
----
$ sudo systemctl start microshift
----
```

Author: done

// * microshift_configuring/microshift-disable-lvms-csi-provider-csi-snapshot.adoc

:_mod-docs-content-type: PROCEDURE
[id="microshift-disabling-lvms-csi-snapshot_{context}"]
= Disabling deployments that run CSI snapshot implementations

Use the following procedure to disable installation of the CSI implementation pods.

[IMPORTANT]
====
This procedure is for users who are defining the configuration file before installing and running {microshift-short}. If {microshift-short} is already started then CSI snapshot implementation will be running. Users must manually remove it by following the uninstallation instructions.
====

[NOTE]
====
{microshift-short} will not delete CSI snapshot implementation pods. You must configure {microshift-short} to disable installation of the CSI snapshot implementation pods during the startup process.
====

.Procedure

. Disable installation of the CSI snapshot controller by entering the `optionalCsiComponents` value under the `storage` section of the {microshift-short} configuration file in `/etc/microshift/config.yaml`:
+
[source,yaml]
----
# ...
  storage: {} <1>
# ...
----
<1> Accepted values are:
* Not defining `optionalCsiComponents`.

> * Specifying `optionalCsiComponents` field with an empty value ([])

Reviewer 3: ```suggestion
* Specifying `optionalCsiComponents` field with an empty value ([])
* Specifying `optionalCsiComponents` field with a single empty string element (`[""]`)
```

This is because of a very irritating behavior in the controller-gen config generation, which we can't work around.

> * Specifying `optionalCsiComponents` field with an empty value ([])

Reviewer 3: ```suggestion
* Specifying `optionalCsiComponents` field with an empty value (`[]`) or a single empty string element (`[""]`)
```
Wrapped the elements with back-ticks to highlight that these are the field values, like the `snapshot-controller` et. al. below. I had to add the alternate value because of a very irritating behavior in the config generator tool we use.

Author: done
